pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+	"github.com/crossplaneio/crossplane/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	managedControllerName   = "managedresource.crossplane.io"
+	managedFinalizerName    = "finalizer." + managedControllerName
+	managedReconcileTimeout = 1 * time.Minute
+
+	defaultManagedShortWait = 30 * time.Second
+	defaultManagedLongWait  = 1 * time.Minute
+)
+
+// Error strings.
+const (
+	errGetManaged    = "cannot get managed resource"
+	errPublishSecret = "cannot publish managed resource connection secret"
+)
+
+// ConnectionDetails created or updated during an operation on an external
+// resource, for example usernames, passwords, endpoints, ports, etc.
+type ConnectionDetails map[string][]byte
+
+// A ManagedPublisher publishes the supplied ConnectionDetails for the supplied
+// Managed resource. ManagedPublishers must handle the case in which the
+// supplied ConnectionDetails are empty.
+type ManagedPublisher interface {
+	Publish(ctx context.Context, mg Managed, c ConnectionDetails) error
+}
+
+// A ManagedCompleter completes the creation of the supplied Managed resource.
+// This often involves adding a finalizer to the object metadata.
+type ManagedCompleter interface {
+	Complete(ctx context.Context, mg Managed) error
+}
+
+// A PublisherChain chains multiple ManagedPublishers.
+type PublisherChain []ManagedPublisher
+
+// Publish calls each ManagedPublisher serially. It returns the first error it
+// encounters, if any.
+func (pc PublisherChain) Publish(ctx context.Context, mg Managed, c ConnectionDetails) error {
+	for _, p := range pc {
+		if err := p.Publish(ctx, mg, c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// An ExternalConnecter produces a new ExternalClient given the supplied
+// Managed resource.
+type ExternalConnecter interface {
+	Connect(ctx context.Context, mg Managed) (ExternalClient, error)
+}
+
+// An ExternalClient manages the lifecycle of an external resource.
+type ExternalClient interface {
+	// Observe the external resource the supplied Managed resource represents,
+	// if any. Observe implementations must not modify the external resource,
+	// but may update the supplied Managed resource to reflect the state of the
+	// external resource.
+	Observe(ctx context.Context, mg Managed) (ExternalObservation, error)
+
+	// Create an external resource per the specifications of the supplied
+	// Managed resource.
+	Create(ctx context.Context, mg Managed) (ExternalCreation, error)
+
+	// Update the external resource represented by the supplied Managed
+	// resource, if necessary. Update implementations must handle the case in
+	// which no update is necessary.
+	Update(ctx context.Context, mg Managed) (ExternalUpdate, error)
+
+	// Finalize the external resource upon deletion of its associated Managed
+	// resource. Finalize implementations may determine whether the external
+	// resource should also be deleted.
+	Finalize(ctx context.Context, mg Managed) error
```

Comment:
I think we do have enough information on whether we want to delete the resource on cloud provider or not through `reclaimPolicy` on `Managed`. We can require the developer to implement a `Delete` function that deletes the object in cloud provider no matter what, but we'd only call it when `g.Spec.ReclaimPolicy == corev1alpha1.ReclaimDelete`, otherwise we'd delete only the `Managed` and not touch the object in cloud provider.

Reply:
What you're describing reflects the current reclaim logic. Per #569 we've discussed changing this logic to more closely reflect Kubernetes PVCs, in which the reclaim policy describes what happens to the managed resource when you delete its bound claim, not what happens to the external resource when you delete its corresponding managed resource. If we do adopt this design we won't have any logic to consider; we'd simply delete the external resource any time the managed resource was deleted.

Either way, we could move the reclaim logic into this reconciler as you suggest and then adapt it once we have a clearer path forward on #569.

Reply:
What do we need to do to obtain a clearer path for the reclaim policy? With the implementation of the shared managed resource reconciler here all reclaim policy for Crossplane will be defined here or left up to the judgement of the creators of managed resource types in other repos.

Reply:
I was expecting something similar to Kubernetes PVCs but then realized that we got that reclaim policy in 2 levels; `claim <> managed` and `managed <> physical resource`. So, my comment assumed that this is the path we're following.

I agree that opting for Kubernetes PVCs model is more practical. The only worry I had about that is our `Managed` resources are in namespaces and namespaces can be just deleted at any time, which would trigger deletion of physical resource that might have been in use for other purposes, or even in other environments(protections around that would block deletion of namespace and I don't think we want that). Knowing that, users might feel they have to allocate the resource solely to that one crossplane, which would limit the use cases for crossplane. Cluster-scoped vs namespace-scoped discussion is here https://github.com/crossplaneio/crossplane/issues/92

I think it'd be easier to do the switch if we make the distinction now and keep the function as only `Delete`, implement the `managed <> physical resource` level reclaimPolicy around where we call that `Delete`. Because two cases could emerge from #92 :
- If we go with cluster-scoped and 1-level reclaim policy, basically k8s PVC model, what we'll need to do is to remove reclaim logic around `Delete` call. No effect on what contributor implemented.
- If we keep it namespaced, well, no change.

But in the current case, if we go with cluster-scoped, `Finalize` would have to become `Delete` in terms of functionality anyway and the name will have to stay as `Finalize` even if it's actually `Delete`. This could require changes in what contributor implemented.

Reply:
I think we can assume with reasonably high confidence that managed resources will soon become cluster scoped, and that using a managed resource without first creating a resource claim will be unsupported, if only by convention.

Given this, I agree that for the moment we should rename `external.Finalize` to `external.Delete` and move the reclaim logic into this controller. We should add a TODO referencing #569 similar to the equivalent comment in the claim reconciler, which will also need updating.

Reply:
Happy to take on the implementation of the reclaim policies!

Reply:
I just left a comment on https://github.com/crossplaneio/crossplane/issues/569

I think what we have right now is that we use `Managed.Spec.ReclaimPolicy` for relation between `Managed` and `Physical Resource`. But in Kubernetes PVC model; `PV.Spec.ReclaimPolicy` is used for relation between `PVC` and `PV`, not `PV` and `Physical Volume`.

I am keeping the behavior as is but I firmly believe we need to use `ReclaimPolicy` for `Claim <> Managed` and `AnotherWordPolicy` for `Managed <> Physical resource`